Fix various standards conformance issues in includes shown by building against musl

ryao opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

### Description
<!--- Describe your changes in detail -->

Fixed includes to use errno.h, poll.h and signal.h instead of their sys/* versions for UNIX System V compliance. Added includes for sys/types.h and unistd.h to cmd/zed/zed_log.c to fix implicit function declarations.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Building with GCC 4.9.4 against musl on Gentoo was noisy due to a mix of preprocessor warnings and compiler warnings.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->

After witnessing the problem, I patched the files, ran `make clean; make` and observed that the problem with noisy includes was gone.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [x] Change has been approved by a ZFS on Linux member.

ryao commented:
Do the buildbots really lack /usr/include/errno.h or is there some package that needs to be installed that is missing? That file is required by the Single UNIX Specification.

loli10K commented:
> Do the buildbots really lack /usr/include/errno.h or is there some package that needs to be installed that is missing? That file is required by the Single UNIX Specification.

If the buildbot were really missing "/usr/include/errno.h" we would not make to the modules compilation:

```
root@debian-8-zfs:/usr/src/zfs# mv /usr/include/errno.h /usr/include/errno.h.bak
root@debian-8-zfs:/usr/src/zfs# make | tail
mkdirp.c:43:19: fatal error: errno.h: No such file or directory
 #include <errno.h>
                   ^
compilation terminated.
make[4]: *** [mkdirp.lo] Error 1
make[3]: *** [all-recursive] Error 1
make[2]: *** [all-recursive] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
  CC       mkdirp.lo
Makefile:683: recipe for target 'mkdirp.lo' failed
make[4]: Leaving directory '/usr/src/zfs/lib/libspl'
Makefile:703: recipe for target 'all-recursive' failed
make[3]: Leaving directory '/usr/src/zfs/lib/libspl'
Makefile:533: recipe for target 'all-recursive' failed
make[2]: Leaving directory '/usr/src/zfs/lib'
Makefile:706: recipe for target 'all-recursive' failed
make[1]: Leaving directory '/usr/src/zfs'
Makefile:574: recipe for target 'all' failed
```

The issue is only in the modules compilation.

loli10K commented on include/sys/nvpair.h:
```diff
@@ -28,7 +28,7 @@
 #include <sys/types.h>
 #include <sys/time.h>
-#include <sys/errno.h>
+#include <errno.h>
```
AFAIK some code is shared among userland and kernel, so instead of "search and replace" `#include <sys/errno.h>` -> `#include <errno.h>` we could do something like this

```c
#ifdef  _KERNEL
#include <linux/errno.h>
#else
#include <errno.h>
#endif  /* _KERNEL */
```

and it should at least compile. I think this is the wrong place to do this kind of trickery though, it would probably better to keep it in the SPL.

behlendorf replied:
The best way to silence these warnings is to add our own `<sys/errno.h>` under `lib/libspl/include/sys/errno.h`.  This path is processed first when searching for headers in the user space build which allows us to avoid these kinds of changes to all the source files.  A simple header which includes <error.h> should do the trick.

```c
#ifndef _LIBSPL_SYS_ERRNO_H
#define _LIBSPL_SYS_ERRNO_H

#include <error.h>

#endif /* _LIBSPL_SYS_ERRNO_H */
```

ryao commented:
@loli10K The kernel modules do not use the system headers.

@behlendorf I like the idea of making sys/*.h headers in libspl. I'll redo this around that.

ryao commented:
@behlendorf I have added headers to libspl. Looking at the build result, there are a few minor warnings that I had missed. Some are due to incorrect header includes in userspace only code, including in the test suite, which I consider to be fine to fix (as all platforms are supposed to support this). Others are due to minor libc differences. I added a patch for one because it broke a zpool command. The remaining errors are far less annoying than the spam that I was seeing previously:

```
kernel.c: In function ‘__dprintf’:
kernel.c:920:25: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
    (void) printf("%u ", (uint_t)pthread_self());
In file included from ../../module/zfs/spa.c:40:0:
../../module/zfs/spa.c: In function ‘spa_deactivate’:
../../module/zfs/spa.c:1257:15: warning: passing argument 1 of ‘zk_thread_join’ makes pointer from integer without a cast
   thread_join(spa->spa_did);
               ^
../../include/sys/zfs_context.h:256:41: note: in definition of macro ‘thread_join’
 #define thread_join(t)   zk_thread_join(t)
                                         ^
../../include/sys/zfs_context.h:264:13: note: expected ‘kt_did_t’ but argument is of type ‘uint64_t’
 extern void zk_thread_join(kt_did_t tid);
             ^
```

ryao commented:
@loli10K If it helps to clarify, your output is from a userland build. The kernel modules do not use the userland headers because the kernel cannot link to the userland libc. The sys/errno.h that we use is actually include/sys/errno.h in the spl repository. The reason this failed is because this code is built both in userspace and the kernel. In the kernel, there is no errno.h (linux/errno.h exists there) and we provide our own sys/errno.h.

loli10K commented:
> your output is from a userland build. The kernel modules do not use the userland headers because the kernel cannot link to the userland libc

@ryao that's kind the issue with your first push right? You stating that the buildbot is missing "/usr/include/errno.h" can't be true because, as i show in my comment, *really* missing that file would break the compilation much earlier.
The issue is that you were trying to `#include <errno.h>` unconditionally for sources used by both userland and kernel, which only breaks the kernel module compilation.

ryao commented:
The buildbot made me realize that glibc includes <sys/poll.h> from <poll.h>, which breaks this. A quick perusal of libc implementations made me realize that klibc and uclibc also do that while Android's bionic does not. I have added exceptions for glibc, klibc and uclibc. While I do not know if we can build against klibc, this will at least mean that builds will not be any harder than they were previously.

ryao commented:
@loli10K I was in a userland mindset from hacking on an initial reimplementation of FUSE when the buildbot complained, so I misread the failure as being in userspace rather than in the kernel. That was a mistake and I see it now.

Also, the buildbot made me realize after I pushed the latest version that glibc breaks the userspace sys/poll.h shim by including <sys/poll.h> from <poll.h>. A quick perusal of libc implementations' headers made me realize that klibc and uclibc also do that while Android's bionic does not. I have added exceptions for glibc, klibc and uclibc. While I do not know if we can build against klibc, this will at least mean that builds will not be any harder on klibc than they were previously. :)

behlendorf commented on lib/libspl/include/sys/errno.h:
```diff
@@ -0,0 +1,32 @@
+ * disallowed by the Single UNIX Specification when building in userspace, so
+ * we implement a dummy header to redirect the include to the proper header.
+ */
+#include <errno.h>
```
We should add add guard macro's for these headers for consistency with the others even if multiple inclusion isn't going to be an issue since all they do is redirect to other headers which do have guard macros.

ryao commented:
@behlendorf I have added the guard macros. I agree that it is good to be consistent with other includes.

behlendorf commented:
Thanks. Now just a few style warnings are left.

ryao commented:
@behlendorf The tab after #define requirement is one that I always forget. It should be fixed now.
